### ISSUE TITLE
Simple example fails with Python 2

### ISSUE THREAD

**Issue by nboullis** [cdent/wsgi-intercept#65]
Hi,

I experienced an issue while using requests and wsgi-intercept to test a WSGI app, with Python 2.
I could simplify the code down to this:
```python
#! /usr/bin/env python

from __future__ import print_function

import wsgiref.simple_server

import requests

from wsgi_intercept.interceptor import RequestsInterceptor

def load_app():
    return wsgiref.simple_server.demo_app

with RequestsInterceptor(load_app, host='www.example.net', port=80):
    r = requests.get('http://www.example.net')
    print(r.text)
```
That code works fine with Python 3 and fails with Python 2.
As I understand it, the problem is that wsgi-intercept builds the `environ` dict, with some keys as unicode strings. It seems to me that all keys for the `environ` dict have to be native strings.
The following patch fixes the issue for my trivial example:
```diff
--- wsgi_intercept/__init__.py	2022-05-13 23:30:34.051461656 +0200
+++ wsgi_intercept/__init__.py	2022-05-13 23:31:03.175107230 +0200
@@ -260,7 +260,7 @@
         else:
             h = k.upper()
             h = h.replace(b'-', b'_')
-            environ['HTTP_' + h.decode('ISO-8859-1')] = v
+            environ['HTTP_' + str(h.decode('ISO-8859-1'))] = v
 
         if debuglevel >= 2:
             print('HEADER:', k, v)
```
However, I’m not 100%-confident that patch is correct.
Any thoughts on this?

**Comment by cdent** [comment]
Thanks for noticing this. I'm surprised it has not come up before. I believe your fix is correct. A few lines above your patch there's a section of code where the value in the dict is coerced into a native str (thus unicode in py3 but not in py2), but that's supposed to be true for the key as well. I'll be pushing up a patch, using your code and test as the basis, in a few minutes. If you're able to confirm that is working as expected, we can get that merged and a new release made.

**Comment by nboullis** [comment]
Hi Chris,
It does work with my trivial patch and with my “real” one.
To be honest, I used your code where the value is coerced into a native string to design my patch.
However, I’m not that confident it is correct for Python 2.
As I understand it, the `make_environ` function receives the request, as bytes, in the `inp` parameter. Here, the headers are supposed to be ISO-8859-1-encoded.
The line
```python
v = str(v.decode('ISO-8859-1'))
```
first decodes the value to Unicode, and the re-encodes it to the system’s default encoding.
But it seems to me that this may be wrong when ISO-8859-1 is not the system’s default encoding.
Shouldn’t you rather do something like
```python
if six.PY3:
    v = v.decode('ISO-8859-1')
```
to only decode the value for Python 3 (like you’re doing, a few lines above, for `method_line`)?
As for the `HTTP_*` headers, I guess you should use something like
```python
if six.PY3:
    h = h.decode('ISO-8859-1')
environ['HTTP_' + h] = v
```
Cheers,
Nicolas

**Comment by cdent** [comment]
> However, I’m not that confident it is correct for Python 2.

PEP 3333 is _very_ confusing on this front, and that confusion compounds in `make_environ` as it is trying to behave like a server. And what happens in the _actual_ headers versus the `environ` gets very messy.

The reason to always decode from ISO-8859-1 is because it allows us to produce an error (decode's default error mode is strict) if any of the headers produced by the requesting user-agent are not valid ISO-8859-1. This allows a test (or anything else) using wsgi-intercept to error out early, indicating that the test (or any of the libs used to create the test) itself is not valid.

It's debatable if this should be the job of the fake server in wsgi-intercept, but it is the pattern that's already been set with the value handling, so may as well also be true for the key handling.

The guarded method line is a special case that comes about as a result of it potentially coming in differently depending on the caller. I can't remember the details but cf94bc1c3f50999e486f7496e1cf92c1a48c40f5 was the original change. In any case the method line is handled differently throughout the stack from the headers.

Yes, totally bewildering, and as a result the tendency in wsgi-intercept is to tweak to be maximally capable and minimally invasive, adjusting as we learn new things or encounter new bugs. In this particular instance you found a way in which things are broken, and the potential fix doesn't appear to add more breakage, so until it does expose more breakage it is probably okay to go with it.

There's more stuff related to this (but more in regard to responses than requests) in 34eb63940db055d15aaf801fac46b6459f04f1cc and #43

**Comment by nboullis** [comment]
Thanks for your answer.
I get your point about the need to check that the headers are valid ISO-8859-1 strings, and about you will to keep your changes minimally invasive.
Cheers,
Nicolas

**Comment by cdent** [comment]
fix has been released as 1.10.0
